zio/csupio/vectorreader.go

```diff
@@ -77,27 +62,29 @@ func (v *VectorReader) Pull(done bool) (vector.Any, error) {
 		v.close()
 		return nil, err
 	}
-	n := int(v.nextObject.Add(1) - 1)
-	if n >= len(v.objects) {
-		return nil, v.close()
+	for {
+		o, err := v.stream.next()
+		if o == nil || err != nil {
+			return nil, err
+		}
+		// XXX using the query context for the metadata filter unnecessarily
+		// pollutes the type context.  We should use the csup local context for
+		// this filtering but this will require a little compiler refactoring to be
+		// able to build runtime expressions that use different type contexts.
+		if v.pushdown == nil || !pruneObject(v.zctx, v.pushdown, o.Metadata()) {
+			return vcache.NewObjectFromCSUP(o).Fetch(v.zctx, v.projection)
+		}
 	}
-	o := v.objects[n]
-	return vcache.NewObjectFromCSUP(o).Fetch(v.zctx, v.projection)
 }
 
-func filterObjects(zctx *super.Context, pruner expr.Evaluator, objects []*csup.Object, projection field.Projection) []*csup.Object {
-	if pruner == nil {
-		return objects
+func pruneObject(zctx *super.Context, pushdown zbuf.Pushdown, m csup.Metadata) bool {
+	metaFilter, metaProjection, err := pushdown.MetaFilter()
+	if err != nil {
+		panic(err)
 	}
```

Comment:
Bit of a shame to recompile this filter every time it's run. Maybe call pushdown.MetaFilter once in NewConcurrentPuller and put the results in VectorReader?

Reply:
@mattnibs pointed out to me that we need separate runtime closures to run concurrently.   Should we recycle compiled expressions in a sync.Pool...?

Reply:
I don't think that's necessary. If you call pushdown.MetaFilter in NewConcurrentPuller and store the results in VectorReader, you'll wind up with one expr.Evaluator per goroutine, which should be sufficient.

Reply:
Oops I saw this after pushing the pool change.  How about we keep the pool so that we can change the way concurrency is done and not have to change it?  I'm find reverting to your suggestion here too.  Easy change.

Reply:
Either's fine with me.